calyx/src/analysis/dominator_map.rs

```diff
@@ -0,0 +1,489 @@
+///   @NODE_ID(9) E;
+/// }
+///
+/// These identifiers are used by the compilation methods [calculate_states_recur]
+/// and [control_exits]
+fn compute_unique_ids(con: &mut ir::Control, mut cur_state: u64) -> u64 {
+    match con {
+        ir::Control::Enable(ir::Enable { attributes, .. })
+        | ir::Control::Invoke(ir::Invoke { attributes, .. }) => {
+            attributes.insert(NODE_ID, cur_state);
+            cur_state + 1
+        }
+        ir::Control::Par(ir::Par {
+            stmts, attributes, ..
+        })
+        | ir::Control::Seq(ir::Seq {
+            stmts, attributes, ..
+        }) => {
+            attributes.insert(NODE_ID, cur_state);
+            cur_state += 1;
+            stmts.iter_mut().for_each(|stmt| {
+                let new_state = compute_unique_ids(stmt, cur_state);
+                cur_state = new_state;
+            });
+            cur_state
+        }
+        ir::Control::If(ir::If {
+            tbranch,
+            fbranch,
+            attributes,
+            ..
+        }) => {
+            attributes.insert(BEGIN_ID, cur_state);
+            cur_state += 1;
+            cur_state = compute_unique_ids(tbranch, cur_state);
+            cur_state = compute_unique_ids(fbranch, cur_state);
+            attributes.insert(END_ID, cur_state);
+            cur_state + 1
+        }
+        ir::Control::While(ir::While {
+            body, attributes, ..
+        }) => {
+            attributes.insert(NODE_ID, cur_state);
+            cur_state += 1;
+            cur_state = compute_unique_ids(body, cur_state);
+            cur_state
+        }
+        ir::Control::Empty(_) => cur_state,
+    }
+}
+
+impl DominatorMap {
+    /// Construct a domination map.
+    pub fn new(control: &mut ir::Control, component_name: String) -> Self {
+        compute_unique_ids(control, 0);
+        let mut exits_map = HashMap::new();
+        Self::build_exit_map(control, &mut exits_map);
+        let mut map = DominatorMap {
+            map: HashMap::new(),
+            exits_map,
+            component_name,
+        };
+        Self::build_map(control, &mut map);
+        map
+    }
+
+    //Builds the domination map by running update_map() until the map
+    //stops changing.
+    fn build_map(main_c: &ir::Control, d_map: &mut DominatorMap) {
+        let mut og_map = d_map.map.clone();
+        let empty_set: HashSet<u64> = HashSet::new();
+        Self::update_map(main_c, 0, &empty_set, d_map);
+        while og_map != d_map.map {
+            og_map = d_map.map.clone();
+            Self::update_map(main_c, 0, &empty_set, d_map);
+        }
+    }
+
+    //Given a control, gets its associated id. For if statments, gets the
+    //beginning id if begin_id is true and end_id if begin_id is false.
+    //Should not be called on empty control
+    //statements or any other statements that don't have an id numbering.
+    fn get_id(c: &ir::Control, begin_id: bool) -> u64 {
+        if let Some(v) = match c {
```

Comment:
Change this into a `let v = match { ... }; v.unwrap_or_else(|| unreachable!(..))` call instead of the `if let` stuff. In general, if the `None` branch is an exception, use `unwrap_or_else`